proposals/003-authentication-api.md

```diff
@@ -0,0 +1,223 @@
+
+## Proposal
+
+### Public API changes
+
+Add the following interface to the `kroxylicious-api` module to allow a `Filter` implementation to opt into consuming authentication outcomes:
+
+```java
+package io.kroxylicious.proxy.filter;
+
+import javax.security.auth.Subject;
+import javax.security.auth.login.LoginException;
+
+// This is a new interface in the api module
+interface ClientSubjectAwareFilter extends Filter {
+    // Called when a client authenticates, or reauthenticates
+    void onClientAuthentication(Subject clientSubject, FilterContext context);
+    // Called when a client fails authentication, or reauthentication
+    default void onClientAuthenticationFailure(LoginException exception, FilterContext context) { 
+    }
+}
+```
+
+Add the following two methods to `FilterContext` to allow SASL terminating `Filter` implementations to propagate their authentication outcomes to upstream filters:
+
+```java
+public interface FilterContext {
+
+    void clientAuthenticationSuccess(Subject subject);
+
+    void clientAuthenticationFailure(LoginException e);
+    
+}
+```
+
+### Runtime changes
+
+1. Change the `kroxylicious-runtime` module so a `Subject` gets instantiated on initial connection.
+2. On successful completion of a TLS handshake, associate a `javax.security.auth.x500.X500Principal` with that `Subject`.
+3. On construction of the filter chain fire an internal message down the pipeline.
+4. Change the `FilterHandler` and `FilterContext` to supply the subject from such a message to the `Filter` implementation.
+
+### An example identity-consuming `Filter`
+
+This section sketches how a Filter implementation would consume client identity:
+
+```java
+    public class ExampleIdentityConsumingFilter implements ProduceRequestFilter, ClientSubjectAwareFilter {
+        private Subject subject;
+
+        public void onClientAuthentication(Subject clientSubject, FilterContext context) {
+            // Store the subject for use later
+            this.subject = clientSubject;
+        }
+
+        @Override
+        public CompletionStage<RequestFilterResult> onProduceRequest(short apiVersion, RequestHeaderData header, ProduceRequestData request, FilterContext context) {
+            if (subject == null) {
+                return null; // TODO return an error reponse, e.g. Errors.SASL_AUTHENTICATION_FAILED, or Errors.UNKNOWN_SERVER_ERROR
+            }
+            else if (authorized(subject)) {
+                // TODO do something and forward the request
+                return null;
+            }
+            else {
+                return null; // TODO return an error reponse: Errors.TOPIC_AUTHORIZATION_FAILED
+            }
+        }
+
+        private static boolean authorized(Subject subject) {
+            // TODO this is not abstracted from the type of Principal
+            //   We really want a model like Kafka's org.apache.kafka.server.authorizer.Authorizer
```

Comment:
I think what you are proposing is good enough.  We can document which Principals the core of Kroxylicious will use.   That will be good enough for 99% of use-cases.

Reply:
If we provide a SaslSniffingFilter as part of this work (I think we should). We need to decide what sort of Principal it will add.

Reply:
marking resolved, specific principal types are now proposed in the latest update.